static/js/favicon.js

```diff
@@ -52,13 +70,45 @@ export function update_favicon(new_message_count, pm_count) {
                 ? `${Math.floor(new_message_count / 1e3 ** pow)}${suffix}`
                 : "∞";
         const count_long = count.length > 2;
+
+        // we need to generate a data url, we cannot just pass a href into
+        // the template because we generate an immutable blob from the
+        // template which would not include the rendered image
+        if (user_settings.realm_icon_as_favicon && page_params.realm_icon_url) {
+            // generate new icon_data_url if necessary
+            if (!current_realm_icon_url || current_realm_icon_url !== page_params.realm_icon_url) {
+                const response = await fetch(page_params.realm_icon_url);
```

Comment:
Waiting for a network request here has some failure modes.

* There could be many updates to the count while the image is still loading. This will generate many image requests, and they will resolve in unclear order, so the count displayed at the end might be wrong.
* The image could fail to load, in which case the count will never be updated at all.

I think we want `update_favicon` to remain synchronous. We want a separate task, triggered by page load or settings updates, that fetches the image, stores the data URL, and does an `update_favicon` with the current counts as of when the image *finished* loading. Then the user will always see an up-to-date count even if the image is stale. This perhaps means the `favicon` module needs to remember the current counts, which would also solve some problems you had elsewhere.

Reply:
if you're concerned about that, can we go one step further and precompute the data url on the backend with PIL or something and make that available via page params (like the commit message talked about)? If we go down that path we should obviously add a DB entry rather than computing at runtime... 

The main reason why I'm interested in that direction is because it would mean that we don't need to wait for a request at all in the normal (`update_favicon`) path. We would still need to add a (`update_realm_icon_data_url`) to handle the case of a new realm icon being uploaded (which would only be called by `server_events_dispatch`), but I suspect this would be simpler... would you say what I'm describing is a reasonable plan?